erofs/erofs.go

```diff
@@ -0,0 +1,744 @@
+// This package is a small go "library" (read: exec wrapper) around the
+// mkfs.erofs binary that provides some useful primitives.
+package erofs
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+	"machinerun.io/atomfs/log"
+	"machinerun.io/atomfs/mount"
+)
+
+var checkZstdSupported sync.Once
+var zstdIsSuspported bool
+
+var exPolInfo struct {
+	once   sync.Once
+	err    error
+	policy *ExtractPolicy
+}
+
+// ExcludePaths represents a list of paths to exclude in a erofs listing.
+// Users should do something like filepath.Walk() over the whole filesystem,
+// calling AddExclude() or AddInclude() based on whether they want to include
+// or exclude a particular file. Note that if e.g. /usr is excluded, then
+// everyting underneath is also implicitly excluded. The
+// AddExclude()/AddInclude() methods do the math to figure out what is the
+// correct set of things to exclude or include based on what paths have been
+// previously included or excluded.
```

Comment:
I'm finding this hard to reason about, and therefore finding it hard to figure out whether AddExclude() is doing the right thing.

I think you're mainly thinking in terms of "I am generating a changeset and therefore want to include only files which have changed", is that right?